File: CPP/token_demo.cpp
/*
 * ExpressionKit Token Demo
 * 
 * This demo showcases the token sequence output functionality that allows
 * for syntax highlighting, expression analysis, and other advanced features.
 */

#include "../ExpressionKit.hpp"
#include <iostream>
#include <iomanip>
#include <unordered_map>
#include <vector>
#include <cmath>

using namespace ExpressionKit;

// Function to print token type name
std::string tokenTypeName(TokenType type) {
    switch (type) {
        case TokenType::NUMBER: return "NUMBER";
        case TokenType::BOOLEAN: return "BOOLEAN";
        case TokenType::IDENTIFIER: return "IDENTIFIER";
        case TokenType::OPERATOR: return "OPERATOR";
        case TokenType::PARENTHESIS: return "PARENTHESIS";
        case TokenType::COMMA: return "COMMA";
        case TokenType::WHITESPACE: return "WHITESPACE";
        case TokenType::UNKNOWN: return "UNKNOWN";
    }
    return "UNKNOWN";
}

// Function to demonstrate token collection for an expression
void demonstrateTokens(const std::string& expression, IEnvironment* environment = nullptr) {
    std::cout << "\n" << std::string(60, '=') << std::endl;
    std::cout << "Expression: " << expression << std::endl;
    std::cout << std::string(60, '=') << std::endl;
    
    try {
        // Collect tokens during evaluation
        std::vector<Token> tokens;
        Value result = ExpressionKit::Eval(expression, environment, &tokens);
        
        std::cout << "Result: " << result.toString() << std::endl;
        std::cout << "\nTokens collected (" << tokens.size() << " total):" << std::endl;
        std::cout << std::left << std::setw(12) << "Type" 
                  << std::setw(8) << "Start" 
                  << std::setw(8) << "Length" 
                  << "Text" << std::endl;
        std::cout << std::string(60, '-') << std::endl;
        
        for (const auto& token : tokens) {
            std::cout << std::left << std::setw(12) << tokenTypeName(token.type)
                      << std::setw(8) << token.start
                      << std::setw(8) << token.length
                      << "\"" << token.text << "\"" << std::endl;
        }
        
        // Demonstrate parsing with tokens (for pre-compilation scenarios)
        std::cout << "\n--- Alternative: Parse with tokens, then execute ---" << std::endl;
        std::vector<Token> parseTokens;
        auto ast = ExpressionKit::Parse(expression, &parseTokens);
        Value parseResult = ast->evaluate(environment);
        
        std::cout << "Parse result: " << parseResult.toString() << std::endl;
        std::cout << "Parse tokens: " << parseTokens.size() << " (same as above)" << std::endl;
        
    } catch (const ExprException& e) {
        std::cout << "Error: " << e.what() << std::endl;
    }
}

// Simple environment for variable demonstration
class DemoBackend : public IEnvironment {
    std::unordered_map<std::string, Value> variables;
    
public:
    DemoBackend() {
        variables["x"] = 10.0;
        variables["y"] = 5.0;
        variables["pi"] = 3.14159;
        variables["isActive"] = true;
        variables["player.health"] = 75.0;
        variables["player.maxHealth"] = 100.0;
    }
    
    Value Get(const std::string& name) override {
        auto it = variables.find(name);
        if (it != variables.end()) return it->second;
        throw ExprException("Variable not found: " + name);
    }
    
    Value Call(const std::string& name, const std::vector<Value>& args) override {
        // First try standard mathematical functions
        Value result;
        if (ExpressionKit::CallStandardFunctions(name, args, result)) {
            return result;
        }
        
        // Custom functions
        if (name == "distance" && args.size() == 4) {
            double x1 = args[0].asNumber(), y1 = args[1].asNumber();
            double x2 = args[2].asNumber(), y2 = args[3].asNumber();
            double dx = x2 - x1, dy = y2 - y1;
            return Value(std::sqrt(dx*dx + dy*dy));
        }
        
        throw ExprException("Function not found: " + name);
    }
};

int main() {
    std::cout << "🚀 ExpressionKit Token Sequence Demo" << std::endl;
    std::cout << "=====================================" << std::endl;
    std::cout << "\nThis demo shows how to collect token sequences during expression" << std::endl;
    std::cout << "parsing for syntax highlighting, analysis, and other advanced features." << std::endl;
    
    // Basic arithmetic expression
    demonstrateTokens("2 + 3 * 4");
    
    // Boolean expression with operators
    demonstrateTokens("true && (false || !true)");
    
    // Complex expression with parentheses
    demonstrateTokens("(10 + 5) * 2 - 3");
    
    // Expression with comparison operators
    demonstrateTokens("5 >= 3 && 10 != 8");
    
    // Mathematical functions
    demonstrateTokens("max(10, 5) + sqrt(16)");
    
    // Now with environment for variables and custom functions
    DemoBackend environment;
    
    // Variable access
    demonstrateTokens("x + y * pi", &environment);
    
    // Complex expression with variables
    demonstrateTokens("player.health / player.maxHealth >= 0.5", &environment);
    
    // Function call with variables
    demonstrateTokens("distance(0, 0, x, y)", &environment);
    
    // Mixed expression with everything
    demonstrateTokens("isActive && (player.health > 50) && max(x, y) >= 5", &environment);
    
    std::cout << "\n" << std::string(60, '=') << std::endl;
    std::cout << "🎯 Use Cases for Token Sequences:" << std::endl;
    std::cout << "• Syntax highlighting in code editors" << std::endl;
    std::cout << "• Expression validation and error reporting" << std::endl;
    std::cout << "• Auto-completion for variables and functions" << std::endl;
    std::cout << "• Expression formatting and pretty-printing" << std::endl;
    std::cout << "• Static analysis and optimization" << std::endl;
    std::cout << "• IDE integration and debugging tools" << std::endl;
    std::cout << std::string(60, '=') << std::endl;
    
    return 0;
}

File: Java/src/main/java/com/expressionkit/examples/TokenDemo.java
/*
 * MIT License
 *
 * Copyright (c) 2025 ExpressionKit Contributors
 */

package com.expressionkit.examples;

import com.expressionkit.*;
import java.util.List;

/**
 * Token analysis demonstration for ExpressionKit Java implementation
 * 
 * This demo shows how to collect and analyze tokens during expression parsing,
 * useful for syntax highlighting, IDE integration, and advanced expression analysis.
 */
public class TokenDemo {

    public static void main(String[] args) {
        System.out.println("🎨 ExpressionKit Java Token Analysis Demo");
        System.out.println("=========================================");
        System.out.println();

        // Test expressions with different token types
        String[] testExpressions = {
            "2 + 3 * 4",
            "true && (5 > 3)",
            "\"hello\" + \" world\"",
            "max(x, sqrt(y))",
            "condition ? value1 : value2",
            "pi * pow(radius, 2)",
            "name == \"ExpressionKit\" and active",
            "!(x < 0) or (y >= 10)",
            "floor(3.7) + ceil(2.1)"
        };

        SimpleEnvironment environment = new SimpleEnvironment();
        environment.set("x", 5.0);
        environment.set("y", 16.0);
        environment.set("pi", Math.PI);
        environment.set("radius", 3.0);
        environment.set("name", "ExpressionKit");
        environment.set("active", true);
        environment.set("condition", false);
        environment.set("value1", 10);
        environment.set("value2", 20);

        for (int i = 0; i < testExpressions.length; i++) {
            String expression = testExpressions[i];
            
            System.out.printf("📝 Expression %d: %s%n", i + 1, expression);
            System.out.println("─".repeat(50));

            try {
                // Parse and evaluate with token collection
                Expression.EvalResult result = Expression.evalWithTokens(expression, environment);
                
                // Show evaluation result
                System.out.printf("✅ Result: %s (%s)%n", 
                    formatValue(result.getValue()), 
                    getTypeName(result.getValue()));
                
                System.out.println();
                System.out.println("🔍 Token Breakdown:");
                
                // Analyze tokens
                List<Token> tokens = result.getTokens();
                analyzeTokens(tokens, expression);
                
                System.out.println();
                showSyntaxHighlighting(tokens, expression);
                
            } catch (ExpressionException e) {
                System.err.println("❌ Error: " + e.getMessage());
            }
            
            System.out.println();
            System.out.println("═".repeat(70));
            System.out.println();
        }

        System.out.println("📊 Token Analysis Summary");
        System.out.println("=========================");
        System.out.println();
        System.out.println("Token Types and Their Uses:");
        System.out.println("  🔢 NUMBER      - Numeric literals (42, 3.14, -2.5)");
        System.out.println("  🔘 BOOLEAN     - Boolean literals (true, false)");
        System.out.println("  📝 STRING      - String literals (\"hello\", \"world\")");
        System.out.println("  🏷️  IDENTIFIER  - Variables and functions (x, sqrt, max)");
        System.out.println("  ⚙️  OPERATOR    - All operators (+, -, ==, &&, etc.)");
        System.out.println("  🎯 PARENTHESIS - Grouping symbols ( )");
        System.out.println("  📍 COMMA       - Function argument separator");
        System.out.println("  ⬜ WHITESPACE  - Spaces and tabs (for formatting)");
        System.out.println();
        System.out.println("💡 Use Cases:");
        System.out.println("  • Syntax highlighting in code editors");
        System.out.println("  • Error reporting with precise locations");
        System.out.println("  • Auto-completion suggestions");
        System.out.println("  • Expression validation before evaluation");
        System.out.println("  • Static analysis and complexity metrics");
        System.out.println("  • Pretty-printing and code formatting");
    }

    private static void analyzeTokens(List<Token> tokens, String expression) {
        int[] typeCounts = new int[TokenType.values().length];
        
        System.out.printf("%-4s %-12s %-20s %-10s %-8s%n", 
            "#", "Type", "Text", "Position", "Length");
        System.out.println("─".repeat(60));
        
        for (int i = 0; i < tokens.size(); i++) {
            Token token = tokens.get(i);
            typeCounts[token.getType().getValue()]++;
            
            String typeColor = getTokenTypeColor(token.getType());
            String resetColor = "\u001B[0m";
            
            System.out.printf("%-4d %s%-12s%s %-20s %-10s %-8d%n",
                i + 1,
                typeColor,
                token.getType(),
                resetColor,
                "\"" + escapeString(token.getText()) + "\"",
                token.getStart() + ":" + (token.getStart() + token.getLength()),
                token.getLength());
        }
        
        System.out.println();
        System.out.println("📈 Token Statistics:");
        for (TokenType type : TokenType.values()) {
            int count = typeCounts[type.getValue()];
            if (count > 0) {
                String typeColor = getTokenTypeColor(type);
                System.out.printf("  %s%s%s: %d%n", 
                    typeColor, type, "\u001B[0m", count);
            }
        }
    }

    private static void showSyntaxHighlighting(List<Token> tokens, String expression) {
        System.out.println("🎨 Syntax Highlighting Preview:");
        
        StringBuilder highlighted = new StringBuilder();
        int lastEnd = 0;
        
        for (Token token : tokens) {
            // Add any gaps (shouldn't happen in practice)
            if (token.getStart() > lastEnd) {
                highlighted.append(expression, lastEnd, token.getStart());
            }
            
            // Add colored token
            String color = getTokenTypeColor(token.getType());
            highlighted.append(color)
                      .append(token.getText())
                      .append("\u001B[0m"); // Reset color
            
            lastEnd = token.getStart() + token.getLength();
        }
        
        // Add any remaining text
        if (lastEnd < expression.length()) {
            highlighted.append(expression.substring(lastEnd));
        }
        
        System.out.println("  " + highlighted.toString());
    }

    private static String getTokenTypeColor(TokenType type) {
        switch (type) {
            case NUMBER: return "\u001B[36m"; // Cyan
            case BOOLEAN: return "\u001B[35m"; // Magenta  
            case STRING: return "\u001B[32m"; // Green
            case IDENTIFIER: return "\u001B[33m"; // Yellow
            case OPERATOR: return "\u001B[31m"; // Red
            case PARENTHESIS: return "\u001B[34m"; // Blue
            case COMMA: return "\u001B[37m"; // White
            case WHITESPACE: return "\u001B[90m"; // Gray
            default: return "\u001B[0m"; // Reset
        }
    }

    private static String formatValue(Value value) {
        try {
            if (value.isNumber()) {
                double num = value.asNumber();
                if (num == Math.floor(num) && !Double.isInfinite(num)) {
                    return String.format("%.0f", num);
                }
                return String.valueOf(num);
            } else if (value.isBoolean()) {
                return String.valueOf(value.asBoolean());
            } else if (value.isString()) {
                return "\"" + value.asString() + "\"";
            }
        } catch (ExpressionException.TypeException e) {
            return "<invalid>";
        }
        return value.toString();
    }

    private static String getTypeName(Value value) {
        if (value.isNumber()) return "Number";
        if (value.isBoolean()) return "Boolean";
        if (value.isString()) return "String";
        return "Unknown";
    }

    private static String escapeString(String str) {
        return str.replace("\\", "\\\\")
                  .replace("\"", "\\\"")
                  .replace("\n", "\\n")
                  .replace("\t", "\\t")
                  .replace("\r", "\\r");
    }
}